src/backend/api/meals.js

```diff
@@ -1,15 +1,306 @@
+      response.json(updatedMeal);
+    }
+  } catch (error) {
+    throw error;
+  }
+});
+
+// delete
+router.delete("/:id", async (request, response) => {
+  try {
+    const meals = await knex("meal").select("*");
+    const inputId = Number(request.params.id);
+    const maxIdOfMeal =Math.max( meals.map(meal => meal.id));
+    if (isNaN(inputId)) {
+      response.send("Id is not number");
+    }
+    else if (inputId > maxIdOfMeal) {
+      response.send(`the largest id is : ${maxIdOfMeal}`)
+    }
+    else {
+      const deletedMeal = meals
+        .where({ id: request.params.id })
+        .delete();
+      response.json(deletedMeal);
+    }
+  } catch (error) {
+    throw error;
+  }
+});
+
+
+//GET api/meals/  Query parameters 
+
+router.get("/", async (request, response) => {
+
+  if ('maxPrice' in request.query) {
+    try {
+      const meals = await knex("meal").select("*");
+      const maxPrice = Number(request.query.maxPrice);
+
+      if (isNaN(maxPrice)) {
+        response.send("not a number")
+      } else {
+        const cheapMeal = meals.filter(meal => meal.price <= maxPrice);
+        if (cheapMeal.length > 0) {
+          response.json(cheapMeal);
+        }
+        else {
+          response.send("no such cheap meal")
+        }
+      }
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  if ('title' in request.query) {
+    try {
+      const meals = await knex("meal").select("*");
+      const inputTitle = request.query.title.toLowerCase();
+      const titleMached = meals.filter(meal => meal.title.toLowerCase().includes(inputTitle));
+
+      if (titleMached.length > 0) {
+        response.json(titleMached);
+      }
+      else {
+        response.send("no meals title macthed here")
+      }
+    } catch (error) {
+      throw error;
+    }
+  }
+
+  if ('createdAfter' in request.query) {
+    try {
+      const meals = await knex("meal").select("*");
+      const dateMacthedMeal = meals.filter(meal => {
+        const mealDate = new Date(meal.created_date).getTime();
+        const searchDate = new Date(request.query.createdAfter).getTime();
+        return mealDate > searchDate;
+
+      })
+      if (dateMacthedMeal.length > 0) {
+        response.json(dateMacthedMeal);
+      } else {
+        response.send("No such meal,Pls check the date")
+      }
+    } catch (error) {
+      throw error;
+
+    }
+  }
+
+  if ('limit' in request.query) {
+    try {
+      const meals = await knex("meal").select("*");
+      const limitNum = Number(request.query.limit);
+
+      if (isNaN(limitNum)) {
+        response.send("not a number")
+      }
+      else if (limitNum > meals.length) {
+        response.send(`there are just ${meals.length} meals`)
+      } else {
+        const limitedMeal = meals.slice(0, limitNum);
+        response.json(limitedMeal);
+      }
+
+    } catch (error) {
+      throw error;
+    }
+  }
+  
+// if ('availableReservations=true' in request.query) {
```

Comment:
You are close! Try it like this:
```
await knex("meal")
      .join("Reservation", { "meal.id": "Reservation.meal_id" })
      .select(
        "meal.id",
        "meal.title",
        "meal.when",
        "meal.max_reservations  as Total_reservations",
        knex.raw(
          'meal.max_reservations - sum(Reservation.number_of_guests) as "No_of_available_reservations"'
        )
      )
      .sum("Reservation.number_of_guests as reservations_booked")
      .having(
        "meal.max_reservations",
        ">",
        "sum('Reservation.number_of_guests')"
      )
      .groupBy("Reservation.meal_id");
```

Reply:
Hi, good morning. I used your way and it is not working.

Reply:
await knex("meal")
.join("reservation", "meal.id", "=", "reservation.meal_id")
      .select(
        "meal.id",
        "title",
        "max_reservations",
        knex.raw("SUM(number_of_guests) AS total_guests"),
        knex.raw(
          '(max_reservations-SUM(number_of_guests)) AS "Available Reservation"'
        )
      )
      .where("max_reservations", ">", "number_of_guests")
      .groupBy("meal_id")
      .having(knex.raw("(max_reservations-SUM(number_of_guests)) > 0"));


This works for me.